Refactor the bundle installation to be generic

renatho opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to https://github.com/Automattic/themes/issues/7498

First part of this refactor: https://github.com/Automattic/wp-calypso/pull/84483

## Proposed Changes

* Refactor the bundle installation to be generic.
  * Everything that happens for WooCommerce should continue having exactly the same behavior.
  * It now should support other bundles.
* Remove Woo Verify Email, which was not used, always protected by a feature flag.
  * PR 1 with the feature: https://github.com/Automattic/wp-calypso/pull/64387
  * PR 2 with the feature: https://github.com/Automattic/wp-calypso/pull/64462
  * Commit: https://github.com/Automattic/wp-calypso/pull/84886/commits/ad225e6fc52d699e6d6725acda1875bd20e3472d
  * Slack conversation: p1701903126481889-slack-C0Q664T29
* Remove unused code from `site-setup-flow`.
  * Commit: https://github.com/Automattic/wp-calypso/pull/84886/commits/d8ea69e997811d0e3c2d4bb1a419796ba967a2fc
  * Slack conversation: p1702302906398119-slack-C0Q664T29

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

The following tests are just to make sure everything continues working as previously. Nothing should change and anything new was introduced. Also, please test other different scenarios that you can think of.

**Flow starting with free plan and selecting a WooCommerce theme**

1. Navigate to `http://calypso.localhost:3000/start`.
2. Select the free plan.
3. On the "Pick a design" step, go to the "store" tab, and select the "Tsubaki" theme (WooCommerce theme).
4. Continue through the flow, upgrading the plan to be able to install the theme.
5. Make sure everything works in the same way as previously (a good way I used to compare was to navigate in wordpress.com in parallel), and when you complete all the steps you are redirected to the WooCommerce home with WC installed.

**Flow starting with business plan and selecting a WooCommerce theme**

1. Navigate to `http://calypso.localhost:3000/start`.
2. Select the business plan.
3. On the "Pick a design" step, go to the "store" tab, and select the "Tsubaki" theme (WooCommerce theme).
4. Continue through the flow.
5. Make sure everything works in the same way as previously (a good way I used to compare was to navigate in wordpress.com in parallel), and when you complete all the steps you are redirected to the WooCommerce home with WC installed.

**Flow starting with free plan and free theme changing to WooCommerce theme later**

1. Navigate to `http://calypso.localhost:3000/start`.
2. Select the free plan.
3. On the "Pick a design" step, select the Twenty Twenty-Four theme.
4. Continue through the flow until redirected to the site.
5. On the site, navigate to Appearance > Themes.
7. Select the "Tsubaki" theme.
8. Upgrade, and continue through the flow.
9. Make sure everything works in the same way as previously (a good way I used to compare was to navigate in wordpress.com in parallel), and when you complete all the steps you are redirected to the WooCommerce home with WC installed.

**Flow starting with business plan and free theme, and changing to WooCommerce theme later**

1. Navigate to `http://calypso.localhost:3000/start`.
2. Select the business plan.
3. On the "Pick a design" step, select the Twenty Twenty-Four theme.
4. Continue through the flow.
5. Make sure everything works in the same way as previously (a good way I used to compare was to navigate in wordpress.com in parallel). Notice that in this case, WooCommerce wasn't installed. Not sure if it's the ideal behavior, but it keeps the current behavior. If confirmed that it's not an intentional behavior, I can open an issue for that.

## Known issues

* When I started developing this PR, I noticed an intermittent error (also reproduced on `trunk` and staging). When I completed the flow with a WooCommerce bundle (`woo-on-plans`), the _"Customize this design"_ button in `http://calypso.localhost:3000/marketplace/thank-you/{DOMAIN}` was always a link to the site editor. But sometimes when clicked, it was redirecting to WooCommerce home (`/wp-admin/admin.php?page=wc-admin`), and others it was just navigating to the site editor without the redirect. I didn't find where this redirect logic lives (I suspect it's not in Calypso), and I couldn't reproduce it anymore, so I'm ignoring it. If you are testing it, and see the same behavior, I can open an issue for that.

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [x] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

renatho committed:
Refactor check for plugins to be generic

renatho committed:
Update bundle settings hook to allow getting settings by theme software

renatho committed:
Move array of plugins to check to the bundle settings

renatho commented on client/my-sites/theme/hooks/use-bundle-settings.tsx:
```diff
@@ -23,18 +26,10 @@ const WooOnPlansIcon = () => (
- */
-const useBundleSettings = ( themeId: string ): BundleSettings | null => {
-	const themeSoftwareSet = useSelector( ( state ) => getThemeSoftwareSet( state, themeId ) );
+const useBundleSettings = ( themeSoftware?: string ): BundleSettingsHookReturn => {
```
It allows to get the settings by theme software or theme (through `useBundleSettingsByTheme`) now.

renatho committed:
Update feature flag to a generic name

renatho committed:
Rename woo confirm to bundle confirm

renatho committed:
Add new hook to get site bundle settings

renatho committed:
Add TODO comment to review later

renatho committed:
Rename woo transfer to bundle transfer

renatho committed:
Make bundle transfer step generic

renatho committed:
Rename woo install plugins to bundle install plugins

renatho committed:
Make install plugins generic

renatho committed:
Translate message that was not translated

renatho committed:
Get software name dynamically from bundle settings

renatho commented on client/landing/stepper/declarative-flow/internals/steps-repository/bundle-install-plugins/index.tsx:
```diff
@@ -17,14 +20,16 @@ export interface FailureInfo {
 		useSelect( ( select ) => select( SITE_STORE ) as SiteSelect, [] );
 	const site = useSite();
-	const softwareSet = 'woo-on-plans';
+	const softwareSet = useSitePluginSlug();
```
In some cases, you will see `softwareSet` vs `pluginSlug`.
They are the same thing. They just have different names because in different places they were called differently, and making everything generic, they are getting mixed in some parts.
I didn't want to rename all functions and variables to not make it even bigger, but it's something that could be done in the future.

renatho committed:
Remove a not necessary hook

renatho committed:
Refactor hooks to be named exports, following the standards of the existing ones

renatho committed:
Use hook to get plugin slug

renatho committed:
Remove Woo Verify Email not used feature

renatho committed:
Reorganize bundle steps array

renatho committed:
Add comments to make the steps more clear

renatho committed:
Move custom back to settings

renatho committed:
Move flow settings to bundleStepsSettings

renatho committed:
Move steps to settings

So only one object from this file needs to be customized for bundles.

renatho committed:
Make end of flow generic

renatho committed:
Make custom functions optional

renatho committed:
Add comments to the custom functions

renatho committed:
Rename the variable to follow the same standard of the other file

renatho committed:
Force type to be a string

renatho commented on client/landing/stepper/declarative-flow/plugin-bundle-data.ts:
```diff
@@ -1,26 +1,100 @@
+};
 
-export default pluginBundleSteps;
+export type BundledPlugin = keyof typeof bundleStepsSettings & string;
```
I don't have much experience with TypeScript, so I'd appreciate help here if any reviewer has a suggestion.
`BundledPlugin` type is currently accepting any `string`.

If I remove the type `BundleStepsSettings` from the `bundleStepsSettings` it works properly, accepting only `"woo-on-plans"`. The same happens with the previous version of this object (`trunk`).

Any idea on how to solve it? The best resource I could find about this is https://stackoverflow.com/questions/51808160/keyof-inferring-string-number-when-key-is-only-a-string

renatho committed:
Add steps as argument to the useStepNavigation

renatho commented on client/landing/stepper/declarative-flow/internals/index.tsx:
```diff
@@ -136,7 +136,11 @@ export const FlowRenderer: React.FC< { flow: Flow } > = ( { flow } ) => {
+	const stepNavigation = flow.useStepNavigation(
+		currentStepRoute,
+		_navigate,
+		flowSteps.map( ( step ) => step.slug )
```
I added the `steps` argument here.

It seems [when it was created](https://github.com/Automattic/wp-calypso/commit/b57297fac00a56cf103f3235e33f6636ddca1f49), the [type already had the steps argument](https://github.com/Automattic/wp-calypso/commit/b57297fac00a56cf103f3235e33f6636ddca1f49#diff-b07941382fbff706bdda39e9b551a1827566a6d90d9ed3f6e286ad7a9e62d02dR25), but it was never used.

renatho committed:
Make the navigation to the next step generic for the submit

renatho committed:
Refactor submit

renatho committed:
Update log and event name

renatho commented on client/landing/stepper/declarative-flow/internals/steps-repository/bundle-confirm/index.tsx:
```diff
@@ -252,7 +255,13 @@ const WooConfirm: Step = function WooCommerceConfirm( { navigation } ) {
-								recordTracksEvent( 'calypso_woocommerce_dashboard_confirm_submit', {
+								let eventName = 'calypso_bundle_dashboard_confirm_submit';
+
+								if ( 'woo-on-plans' === pluginSlug ) {
```
This logic was applied in some different places as part of this commit: https://github.com/Automattic/wp-calypso/pull/84886/commits/4d3209580dd55d0173de427456027482704f11f9

Do you think it makes sense? So we keep the same event and log when it's the `woo-on-plans`, but we use a generic name for the others. Or maybe should we register one event and log for `woo-on-plans`, and always register the generic one (having 2 registers for `woo-on-plans`)?

Personally, I thought the ideal would be to just rename it for every case (the `calypso_woocommerce_dashboard_confirm_submit` wouldn't exist anymore), but I'm afraid that it could break existing reports, and so on.

Also, should I document this somewhere?

dsas replied:
There's a tool for documenting new tracks events, more info: PCYsg-mUj-p2

dsas replied:
I'd be tempted to have both events, one for `woo-on-plans` specifically, so as not to break any existing reports (though it doesn't look oft used tbh) and the generic one that's more future-proof

renatho replied:
Done in: https://github.com/Automattic/wp-calypso/pull/84886/commits/59878f27314cabf956c55e618cf751429dfaf91c. I wasn't very sure about the logstash, so I did the same there, just in case.

I'll take a look about the documentation! Thank you!

renatho replied:
Events are also registered.
The previous ones (WooCommerce events) weren't registered, and I didn't register them because we probably don't want them to be used anymore.

https://github.com/Automattic/tracks-events-registration/pull/2054
https://github.com/Automattic/tracks-events-registration/pull/2055

renatho commented on client/my-sites/theme/hooks/use-bundle-settings.tsx:
```diff
@@ -23,22 +27,15 @@ const WooOnPlansIcon = () => (
- * Hook to get the bundle settings for a given theme.
- * If the theme doesn't have a sotfware set defined, it returns `null`.
- */
-const useBundleSettings = ( themeId: string ): BundleSettings | null => {
```
Now we have hooks to get the settings by software or theme (through `useBundleSettingsByTheme`) now.

renatho committed:
Update import to the new function

renatho committed:
Remove unused code

This logic is part of plugin-bundle-flow.

renatho committed:
Move the styles to the file

Following the approach applied in other files

renatho commented on client/landing/stepper/declarative-flow/internals/steps-repository/bundle-confirm/index.tsx:
```diff
@@ -11,21 +11,41 @@ import FormattedHeader from 'calypso/components/formatted-header';
 import type { TransferEligibilityError } from '@automattic/data-stores/src/automated-transfer-eligibility/types';
 import './style.scss';
 
+const ActionSection = styled.div`
```
I moved these styles directly to this final instead of importing it from `woocommerce-install` to not have it tied to woocommerce components, and following the approach used in other components. Ideally, in the future, it could be in a central place and replace all the occurrences.

renatho committed:
Remove comment

Keeping it as a customization inside the step, as exception.

renatho committed:
Merge branch 'trunk' of github.com:Automattic/wp-calypso into change/make-bundle-installation-generic

dsas commented on client/landing/stepper/declarative-flow/internals/steps-repository/bundle-confirm/index.tsx:
```diff
@@ -252,7 +273,13 @@ const WooConfirm: Step = function WooCommerceConfirm( { navigation } ) {
+								}
+
+								recordTracksEvent( eventName, {
 									site: wpcomDomain,
```
it'd be good to add a property here for the bundle slug, that way we can create bundle specific funnels and stats later if necessary

renatho replied:
Good point! Added in https://github.com/Automattic/wp-calypso/pull/84886/commits/59878f27314cabf956c55e618cf751429dfaf91c.

dsas reviewed (commented):
Wow, amazing work @renatho, thank you for such a thorough job.

There's also the `/setup/ecommerce` entrypoint, I tested it and that also continues to work 👍

okmttdhr commented on client/components/theme-upgrade-modal/index.tsx:
```diff
@@ -73,7 +73,7 @@ export const ThemeUpgradeModal = ( {
 	const isExternallyManaged = useSelector( ( state ) => isExternallyManagedTheme( state, slug ) );
 
-	const bundleSettings = useBundleSettings( slug );
+	const bundleSettings = useBundleSettingsByTheme( slug );
```
I applied your PR in https://github.com/Automattic/wp-calypso/pull/85013, and I changed to use `useBundleSettings`, as I wanted to keep `ThemeUpgradeModal` Calypso-State-free. Could you apply it here as well? Thank you!

- https://github.com/Automattic/wp-calypso/pull/85013#discussion_r1424374607
- https://github.com/Automattic/wp-calypso/pull/85013/files#diff-37aa85681c6036ea769d4a6169a3d623a36082cdba5dcbcfc48bca50214fb4b6
```suggestion
	const bundleSettings = useBundleSettings( slug );
```

renatho replied:
Hey @okmttdhr! Of course!

Here is the change: https://github.com/Automattic/wp-calypso/pull/84886/commits/15cee75c1088ad98347d014316eedeb7574aee77

okmttdhr replied:
LGTM :) Thank you!

renatho committed:
Update logic to upgrade site when installing free theme with bundle

renatho commented on client/my-sites/theme/main.jsx:
```diff
@@ -971,7 +972,11 @@ class ThemeSheet extends Component {
 		} else if ( isLoggedIn && siteId ) {
-			if ( isPremium && ! isThemePurchased && ! isExternallyManagedTheme ) {
+			if (
+				( isPremium || isBundledSoftwareSet ) &&
```
I changed it as part of this commit https://github.com/Automattic/wp-calypso/pull/84886/commits/cd6f5da9c78e746603edd34085c31e0a36b87b88, so free themes will also ask for upgrade when it has a bundle.

~~To test it, you can checkout the branch from this PR: https://github.com/Automattic/wp-calypso/pull/85309, and try to install the "Course" theme on a site with the free plan.~~

dsas replied:
🤔 I'm afraid I'm still getting 'Activate' rather than "upgrade to activate' in the three dot menu on a Free site, whether on this branch or #85309 PR.

<img width="803" alt="Screenshot 2023-12-18 at 15 15 42" src="https://github.com/Automattic/wp-calypso/assets/93301/0033cb27-c803-4704-a4b8-78f5ec9bae95">

dsas replied:
oh hmm, you might get different results in development vs wpcalypso (the calypso live links in pr comments) due to some WIP refactoring we have going on.

What happens if you change `config/development.json` to set `themes/tiers: false` in your local dev env? Does this code work?

renatho replied:
Good catch! I completely missed this one.

The original commit was supposed to fix the button of the following page for free themes with bundle:

<img width="922" alt="Screenshot 2023-12-18 at 17 39 12" src="https://github.com/Automattic/wp-calypso/assets/876340/be74d8c9-7969-4a5d-9f16-07c662edd996">

But while checking it to fix the dropdown link, I noticed that the `isThemePurchased` wasn't correct for this situation, so I also changed that a bit to check if the site has the feature to install plugins.

Here is the new commit: https://github.com/Automattic/wp-calypso/pull/84886/commits/c08ac456da7428c2d3f831d140d3a6654f312cae

Just notice that it doesn't "fix" (or change?) the behavior we talked about in p1700684964931779/1700605581.879429-slack-C048CUFRGFQ, that if the site was already upgraded, it only installs the theme.

I'm not super confident about this change, mainly about the `client/my-sites/themes/theme-options.js` that has many checks in the functions `hideForTheme` for different situations. I'm a little afraid I'm missing something.

renatho replied:
> What happens if you change config/development.json to set themes/tiers: false in your local dev env? Does this code work?

I tested it with the feature flag enabled and disabled and it worked for me.

I also tested with Course (free and bundled), Tsubaki (premium and bundled), a premium theme, and a free theme.

renatho replied:
> To test it, you can checkout the branch from this PR: https://github.com/Automattic/wp-calypso/pull/85309, and try to install the "Course" theme on a site with the free plan.

BTW, I think it's fine to test this in this branch directly. You just need to go to that branch (already updated with the last commit) if you want to see all the correct messages and complete the whole flow. 😉

dsas replied:
> I'm not super confident about this change, mainly about the client/my-sites/themes/theme-options.js that has many checks in the functions hideForTheme for different situations. I'm a little afraid I'm missing something.

Yeah, we don't like it either! We'd dearly like to get rid of it when we can.

renatho committed:
Merge branch 'trunk' of github.com:Automattic/wp-calypso into change/make-bundle-installation-generic

renatho committed:
Make `ThemeUpgradeModal` Calypso-State-free

Update to match changes from conflicting PR. See #85013 and https://github.com/Automattic/wp-calypso/pull/84886#discussion_r1426350658.

Co-authored-by: okmttdhr <okmttdhr@users.noreply.github.com>

renatho committed:
Add event for all the cases, and keep the current one for backward compatibility

renatho committed:
Merge branch 'trunk' of github.com:Automattic/wp-calypso into change/make-bundle-installation-generic

renatho committed:
Merge branch 'trunk' of github.com:Automattic/wp-calypso into change/make-bundle-installation-generic

renatho committed:
Fix case when endFlow is not set

renatho committed:
Merge branch 'trunk' of github.com:Automattic/wp-calypso into change/make-bundle-installation-generic

renatho commented on client/landing/stepper/declarative-flow/plugin-bundle-flow.ts:
```diff
@@ -209,14 +209,15 @@ const pluginBundleFlow: Flow = {
 						exitFlow,
 					} );
-					if ( false !== endReturn ) {
+					if ( settings?.endFlow && false !== endReturn ) {
```
While testing it with the Sensei bundle (https://github.com/Automattic/wp-calypso/pull/85309), I noticed an issue here. Fixed as part of this commit: https://github.com/Automattic/wp-calypso/pull/84886/commits/b5fa4e03ae80ba0cda44c7a7c7e932b0e1a510be

It should return earlier only **if the `endFlow`** exists and it doesn't return `false`.

renatho committed:
Fix logic to display activate or upgrade button

renatho committed:
Merge branch 'trunk' of github.com:Automattic/wp-calypso into change/make-bundle-installation-generic

renatho committed:
Merge branch 'trunk' of github.com:Automattic/wp-calypso into change/make-bundle-installation-generic

renatho committed:
Merge branch 'trunk' of github.com:Automattic/wp-calypso into change/make-bundle-installation-generic

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/10448936).

ilyasfoo commented on client/my-sites/theme/main.jsx:
```diff
@@ -1053,9 +1056,10 @@ class ThemeSheet extends Component {
+				( ( ! isThemeAllowedOnSite || isPremium ) &&
+					! isThemePurchased &&
+					! isExternallyManagedTheme ) ||
+				( isBundledSoftwareSet && ! canInstallPlugins )
```
Hey, @renatho!

This particular change may have unintentionally caused a bug in Woo Express, where themes like `tsubaki` and `tazza` now requires an upgrade. This is because a free trial site can't activate plugins, and I believe `woocommerce` themes are marked as "bundled software set".

May I know what was the intended themes to be matched with this logic?

To reproduce:

1. Create a Woo Express free trial site by going to https://woo.com/start and follow through the NUX process
2. Go to Appearance > Themes
3. Search for `tazza`
4. Select Tazza theme
5. Observe you can't activate it now

<img width="806" alt="image" src="https://github.com/Automattic/wp-calypso/assets/3747241/1ad399e9-7dce-4e7a-bf05-6d794909be78">

ilyasfoo replied:
I've made a PR to fix this, https://github.com/Automattic/wp-calypso/pull/85890.

Appreciate if you can take a look as well, @renatho!

renatho replied:
@ilyasfoo

Thank you for catching it and quickly raising a PR with the fix!

> May I know what was the intended themes to be matched with this logic?

It was supposed to match the paid themes (previous logic) + bundled themes. But I was assuming that any bundled theme would need to be upgraded to a plan where the user can install plugins because it will install a plugin along with the theme. But I missed this use case that you mentioned.

> I've made a PR to fix this, https://github.com/Automattic/wp-calypso/pull/85890.

I'll take a look now.

renatho replied:
> May I know what was the intended themes to be matched with this logic?

The new theme we are considering outside of WooCommerce is the Course theme, which installs the Sensei LMS.
